File: backend-python/app/seed.py
"""Loads the datasets described in contracts/SEED.md.

The starter only has users and teams. When you add projects and tasks, extend
`load_small` and `load_large` so the rest of the seed loads too.
"""

import json
from datetime import datetime
from pathlib import Path

from sqlalchemy import insert
from sqlalchemy.orm import Session

from app.auth import hash_password
from app.db import Base
from app.models import Team, TeamMember, User

SEED_FILE = Path(__file__).resolve().parents[2] / "contracts" / "seed.json"

LARGE_USERS = 500
LARGE_TEAMS = 40


def _ts(value: str) -> datetime:
    return datetime.fromisoformat(value.replace("Z", "+00:00"))


def reset(db: Session, size: str = "small") -> None:
    bind = db.get_bind()
    db.close()
    Base.metadata.drop_all(bind)
    Base.metadata.create_all(bind)
    load_small(db)
    if size == "large":
        load_large(db)
    db.commit()


def load_small(db: Session) -> None:
    seed = json.loads(SEED_FILE.read_text(encoding="utf-8"))
    pw = hash_password(seed["password"])  # one hash reused — scrypt is slow on purpose
    db.execute(
        insert(User),
        [
            {
                "id": u["id"],
                "email": u["email"],
                "first_name": u["firstName"],
                "last_name": u["lastName"],
                "display_name": u["displayName"],
                "role": u["role"],
                "password_hash": pw,
                "created_at": _ts(u["createdAt"]),
                "updated_at": _ts(u["updatedAt"]),
            }
            for u in seed["users"]
        ],
    )
    db.execute(
        insert(Team),
        [
            {
                "id": t["id"],
                "name": t["name"],
                "description": t["description"],
                "created_at": _ts(t["createdAt"]),
                "updated_at": _ts(t["updatedAt"]),
            }
            for t in seed["teams"]
        ],
    )
    db.execute(
        insert(TeamMember),
        [
            {"team_id": m["teamId"], "user_id": m["userId"], "role": m["role"]}
            for m in seed["teamMembers"]
        ],
    )
    db.flush()


def load_large(db: Session) -> None:
    seed = json.loads(SEED_FILE.read_text(encoding="utf-8"))
    pw = hash_password(seed["password"])
    user_base = max(u["id"] for u in seed["users"])
    team_base = max(t["id"] for t in seed["teams"])
    now = _ts("2025-01-01T00:00:00Z")

    db.execute(
        insert(User),
        [
            {
                "id": user_base + n,
                "email": f"user{n}@example.com",
                "first_name": "User",
                "last_name": str(n),
                "display_name": f"User {n}",
                "role": "member",
                "password_hash": pw,
                "created_at": now,
                "updated_at": now,
            }
            for n in range(1, LARGE_USERS + 1)
        ],
    )
    db.execute(
        insert(Team),
        [
            {
                "id": team_base + n,
                "name": f"Team {n:03}",
                "description": "",
                "created_at": now,
                "updated_at": now,
            }
            for n in range(1, LARGE_TEAMS + 1)
        ],
    )
    members = []
    for n in range(1, LARGE_TEAMS + 1):
        users = [u for u in range(1, LARGE_USERS + 1) if u % LARGE_TEAMS == n % LARGE_TEAMS]
        for i, u in enumerate(users):
            members.append(
                {
                    "team_id": team_base + n,
                    "user_id": user_base + u,
                    "role": "lead" if i == 0 else "member",
                }
            )
    db.execute(insert(TeamMember), members)
    db.flush()

File: backend-node/src/seed.ts
/**
 * Loads the datasets described in contracts/SEED.md.
 *
 * The starter only has users and teams. When you add projects and tasks, extend
 * `loadSmall` and `loadLarge` so the rest of the seed loads too.
 */
import { readFileSync } from "node:fs";

import { hashPassword } from "./auth.js";
import { createTables, db, dropTables } from "./db/client.js";

type SeedFile = {
  password: string;
  users: {
    id: number;
    email: string;
    firstName: string;
    lastName: string;
    displayName: string;
    role: "admin" | "member";
    createdAt: string;
    updatedAt: string;
  }[];
  teams: { id: number; name: string; description: string; createdAt: string; updatedAt: string }[];
  teamMembers: { teamId: number; userId: number; role: "lead" | "member" }[];
};

const SEED_FILE = new URL("../../contracts/seed.json", import.meta.url);
const LARGE_USERS = 500;
const LARGE_TEAMS = 40;

function readSeed(): SeedFile {
  return JSON.parse(readFileSync(SEED_FILE, "utf8")) as SeedFile;
}

type NewUser = SeedFile["users"][number] & { passwordHash: string };

function insertUser(u: NewUser) {
  db.run(
    `INSERT INTO users (id, email, first_name, last_name, display_name, role, password_hash, created_at, updated_at)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
    u.id, u.email, u.firstName, u.lastName, u.displayName, u.role, u.passwordHash, u.createdAt, u.updatedAt,
  );
}

function insertTeam(t: SeedFile["teams"][number]) {
  db.run(
    "INSERT INTO teams (id, name, description, created_at, updated_at) VALUES (?, ?, ?, ?, ?)",
    t.id, t.name, t.description, t.createdAt, t.updatedAt,
  );
}

function insertTeamMember(m: SeedFile["teamMembers"][number]) {
  db.run("INSERT INTO team_members (team_id, user_id, role) VALUES (?, ?, ?)", m.teamId, m.userId, m.role);
}

export function reset(size: "small" | "large" = "small"): void {
  dropTables();
  createTables();
  db.transaction(() => {
    loadSmall();
    if (size === "large") loadLarge();
  });
}

export function loadSmall(): void {
  const seed = readSeed();
  const pw = hashPassword(seed.password); // one hash reused — scrypt is slow on purpose
  for (const u of seed.users) insertUser({ ...u, passwordHash: pw });
  for (const t of seed.teams) insertTeam(t);
  for (const m of seed.teamMembers) insertTeamMember(m);
}

export function loadLarge(): void {
  const seed = readSeed();
  const pw = hashPassword(seed.password);
  const userBase = Math.max(...seed.users.map((u) => u.id));
  const teamBase = Math.max(...seed.teams.map((t) => t.id));
  const now = "2025-01-01T00:00:00Z";

  for (let n = 1; n <= LARGE_USERS; n++) {
    insertUser({
      id: userBase + n,
      email: `user${n}@example.com`,
      firstName: "User",
      lastName: String(n),
      displayName: `User ${n}`,
      role: "member" as const,
      passwordHash: pw,
      createdAt: now,
      updatedAt: now,
    });
  }

  for (let n = 1; n <= LARGE_TEAMS; n++) {
    insertTeam({
      id: teamBase + n,
      name: `Team ${String(n).padStart(3, "0")}`,
      description: "",
      createdAt: now,
      updatedAt: now,
    });
  }

  for (let n = 1; n <= LARGE_TEAMS; n++) {
    let first = true;
    for (let u = 1; u <= LARGE_USERS; u++) {
      if (u % LARGE_TEAMS !== n % LARGE_TEAMS) continue;
      insertTeamMember({ teamId: teamBase + n, userId: userBase + u, role: first ? "lead" : "member" });
      first = false;
    }
  }
}
